File: backend/app/schemas/communication.py
import uuid
from datetime import datetime
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.checklist import QualityLevel, Substrate
from app.models.communication import CommunicationCategory
from app.models.risk import RiskSeverity
from app.schemas.risk import RiskSourceFindingRead


class CommunicationApplicationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    inspection_id: uuid.UUID
    phrase_code: str
    phrase_version: int
    category: CommunicationCategory
    priority: int
    phrase_key: str
    why_key: str | None
    seed_key: str | None
    source_kind: str
    source_signature: str
    is_active: bool
    resolved_at: datetime | None
    position: int | None
    created_at: datetime
    updated_at: datetime


class CommunicationApplicationListResponse(BaseModel):
    items: list[CommunicationApplicationRead]
    total: int


class CommunicationRiskSource(BaseModel):
    kind: Literal["RISK"] = "RISK"
    risk_id: uuid.UUID | None
    risk_code: str | None
    severity: RiskSeverity | None
    risk_is_active: bool | None
    source_findings: list[RiskSourceFindingRead] = []


class CommunicationFindingSnapshot(BaseModel):
    finding_id: uuid.UUID
    label_key: str | None
    value_snapshot: dict | None
    is_active: bool
    position: int | None


class CommunicationFindingSource(BaseModel):
    kind: Literal["FINDING"] = "FINDING"
    finding_key: str
    findings: list[CommunicationFindingSnapshot] = []


class CommunicationQualitySource(BaseModel):
    kind: Literal["QUALITY"] = "QUALITY"
    substrate: Substrate | None
    quality_level: QualityLevel | None


CommunicationSource = Annotated[
    CommunicationRiskSource
    | CommunicationFindingSource
    | CommunicationQualitySource,
    Field(discriminator="kind"),
]


class CommunicationApplicationDetailRead(CommunicationApplicationRead):
    source: CommunicationSource

File: frontend/src/types/communication.ts
/**
 * Materialized client-communication applications of a COMPLETED inspection
 * (Stage 8 "Co powiedzieć klientowi"). DTOs mirror backend schemas; all
 * selection/versioning/identity decisions stay on the backend.
 */

import { RiskSeverityValue, RiskSourceFindingRead } from './risk';

export type CommunicationCategoryValue =
  | 'EXPLAIN_CONDITION'
  | 'EXPLAIN_CONSEQUENCE'
  | 'RECOMMEND_PREPARATION'
  | 'REQUIRE_CLIENT_DECISION'
  | 'SCOPE_CLARIFICATION'
  | 'QUALITY_EXPECTATION'
  | 'DOCUMENT_AGREEMENT'
  | 'GENERAL';

export type CommunicationStatusValue = 'active' | 'resolved' | 'all';

export type CommunicationSourceKind = 'RISK' | 'FINDING' | 'QUALITY';

export interface CommunicationApplicationRead {
  id: string;
  inspection_id: string;
  phrase_code: string;
  phrase_version: number;
  category: CommunicationCategoryValue;
  priority: number;
  phrase_key: string;
  why_key: string | null;
  seed_key: string | null;
  source_kind: CommunicationSourceKind;
  source_signature: string;
  is_active: boolean;
  resolved_at: string | null;
  position: number | null;
  created_at: string;
  updated_at: string;
}

export interface CommunicationRiskSource {
  kind: 'RISK';
  risk_id: string | null;
  risk_code: string | null;
  severity: RiskSeverityValue | null;
  risk_is_active: boolean | null;
  source_findings: RiskSourceFindingRead[];
}

export interface CommunicationFindingSnapshot {
  finding_id: string;
  label_key: string | null;
  value_snapshot: Record<string, unknown> | null;
  is_active: boolean;
  position: number | null;
}

export interface CommunicationFindingSource {
  kind: 'FINDING';
  finding_key: string;
  findings: CommunicationFindingSnapshot[];
}

export interface CommunicationQualitySource {
  kind: 'QUALITY';
  substrate: string | null;
  quality_level: string | null;
}

export type CommunicationSource =
  | CommunicationRiskSource
  | CommunicationFindingSource
  | CommunicationQualitySource;

export interface CommunicationApplicationDetailRead
  extends CommunicationApplicationRead {
  source: CommunicationSource;
}

export interface CommunicationListResponse {
  items: CommunicationApplicationRead[];
  total: number;
}